### ISSUE TITLE
[RFC]: MOSS-TTS-Local-v1.5 Serving Optimization

### ISSUE THREAD

**Issue by gcanlin** [vllm-project/vllm-omni#4676]
# RFC: MOSS-TTS-Local-v1.5 Serving Optimization

## Background

MOSS-TTS-Local-v1.5 uses a Global Transformer + Local Transformer architecture. The Global Transformer is a Qwen3-style backbone that models text, reference-audio codes, and previous-frame audio feedback; the Local Transformer expands each global hidden state into one acoustic frame.

Each acoustic frame contains 12 RVQ codebooks, and codebooks inside the frame are generated sequentially. This means every audio frame requires not only one backbone step, but also a local continue/stop decision and 12 codebook sampling steps.

MOSS-Audio-Tokenizer-v2 handles both reference-audio encoding and waveform decoding, targets 48 kHz stereo output, and supports native incremental decoding. This gives MOSS Local high-quality audio and a natural streaming path, but it also makes the serving path heavier than many 24 kHz mono or lightweight-vocoder TTS models.

The current vLLM-Omni MOSS Local path is functionally usable, but it does not yet fully use batching, GPU-resident state, CUDA Graph, tensor payloads, or codec streaming. The goal is to reduce RTF, restore true low-TTFP streaming, and make high-concurrency behavior observable, recoverable, and tunable.

## Goals

- Reduce single-request and concurrent RTF by removing Python hot-path work, CPU syncs, small-kernel launches, and repeated tensor allocations.
- Let Stage 1 consume codec frames before Stage 0 finishes the full utterance, reducing TTFP.
- Preserve offline quality, stop semantics, and voice-clone behavior while adding batched, graphed, and streaming fast paths.
- Keep the optimization path compatible with vLLM-Omni multi-stage serving, benchmark, stats, and the OpenAI-compatible speech API.

### Model Initial Support

- [x] https://github.com/vllm-project/vllm-omni/pull/4664
- [x] https://github.com/vllm-project/vllm-omni/pull/3420

### Reference Audio Preprocessing

- [ ] Add a content-addressed cache for reference-audio codes to avoid repeatedly encoding the same speaker audio. Cache values should be stored on CPU in a compact dtype and returned as clones to avoid accidental mutation by downstream code.

- [ ] Add single-flight for the same uncached reference audio so concurrent requests trigger only one real encoding job. This prevents bursts of voice-clone traffic from overloading the codec encoder with duplicate work.

- [ ] Batch reference-audio encoding with a small wait window. The MOSS codec encoder is heavy enough that modest batching can improve GPU utilization, while per-item failures must still be isolated.

### AR Decode State and Local Transformer

- [x] Support compile for local transformer https://github.com/vllm-project/vllm-omni/pull/5530
- [x] Support CUDA Graph for local transformer #5197
- [x] Batch local-transformer frame decode across active requests. The 12 codebooks still keep their within-frame sequential dependency, but each codebook step should run over the active batch. #5530
- [ ] Reuse local-transformer KV within a frame to avoid re-prefilling the current prefix for every codebook. This must preserve MOSS Local's GPT-J/interleaved RoPE semantics and should be validated against the current eager path for logits and codes.

### Streaming Codec and Vocoder

- [x] Use the native MOSS-Audio-Tokenizer-v2 streaming session. Each streaming request should acquire a slot, decode waveform incrementally by chunk, and release the slot on finish, error, or abort. @gcanlin https://github.com/vllm-project/vllm-omni/pull/4804

- [x] Streaming slot exhaustion must not crash the request. The system should queue, fall back to offline decode, or explicitly reject new requests based on configuration, and the behavior must be observable. @gcanlin https://github.com/vllm-project/vllm-omni/pull/4804

- [x] Add coalesced vocoder steps so when one request is due for decoding, other ready or near-ready slots can join the same codec call. All participating slots should use the same frame count plus an execution mask to amortize codec forward cost. https://github.com/vllm-project/vllm-omni/pull/4804

- [x] Capture common streaming vocoder chunk sizes with CUDA Graph. The codec's internal streaming state must be updated in place so graph replay sees stable buffer addresses. https://github.com/vllm-project/vllm-omni/pull/4929

- [x] Make vocoder streaming batch dynamic and catch the dimension of batch #5235 
 
### GPU-to-GPU Transfer

- [ ] Evaluate CUDA IPC for same-GPU cross-process tensor stream chunks. This should come after delta tensor payloads are implemented, otherwise the benefit for tiny frame rows may be hidden by Python and CPU-path overhead.

- [ ] Evaluate NCCL or another D2D relay for cross-GPU Stage0/Stage1 deployment. This should become a default path only when codec chunk tensors are large enough and CPU relay becomes a visible bottleneck.

## Validation

Correctness validation should cover offline/streaming audio duration, WER/CER, speaker similarity, chunk continuity, stop frames, max-frame handling, abort cleanup, and slot release. Performance validation should report RTF, TTFP, E2E latency, request throughput, audio throughput, graph hit rate, cache hit rate, and per-stage timing by warm state, concurrency level, and stream/non-stream mode.

## CC List

@linyueqian @Sy0307 @hsliuustc0106

**Comment by BeatSeat** [comment]
Should we consider aligning with https://github.com/vllm-project/vllm-omni/issues/4571 ?

**Comment by gcanlin** [comment]
> Should we consider aligning with [#4571](https://github.com/vllm-project/vllm-omni/issues/4571) ?

Yes. It should be aligned because MOSS-TTS-Local also have the local sub-model.

But I'm also considering whether we can refer to the mode of Qwen3-Omni, which doesn't use the inner dependent graph manager. Instead, it's using vLLM native cudagraph wrapper.

**Comment by Sy0307** [comment]
I will close https://github.com/vllm-project/vllm-omni/pull/4686 as I have no bandwidth. Welcome anyone who wants to pick it up and continue optimization work. #4686 solve most performance bottleneck. Someone else can build better performance base on it or solve some disadvantages of it. cc @gcanlin

**Comment by BruceLoveDecimal** [comment]
Hi @gcanlin ,If no one else is working on this and it's open for contribute, I'd like to pick up the **Reference Audio Preprocessing** section. 

After going through the current code, part of it is already in place, so the remaining work should be fairly small:
- **Content-addressed cache** — mostly exists
- **Single-flight** — missing
- **Batched encoding with a small wait window** — missing

**Comment by wendavid552** [comment]
I'd like to pick the AR decode part.

**Comment by gcanlin** [comment]
## Performance Summary

| Configuration | Concurrency / Batch | Requests | Throughput audio duration/s | Mean TTFP / ms | Mean RTF | Mean E2E / ms | Mean WER |
|---|---:|---:|---:|---:|---:|---:|---:|
| single replica (1x H20 GPUs) | 1 | 50 | 4.75 | 92.43 | 0.21 | 879.57 | 0.0298 |
| single replica (1x H20 GPUs) | 8 | 100 | 11.69 | 873.22 | 0.70 | 2885.58 | 0.0298 |
| single replica (1x H20 GPUs) | 16 | 100 | 6.36 | 2978.31 | 2.31 | 10205.36 | 0.0455 |
| single replica (1x H20 GPUs) | 32 | 200 | 8.73 | 3864.10 | 3.07 | 13529.47 | 0.0364 |
| multi-replica (8x H20 GPUs) | 16 | 100 | 14.87 | 306.02 | 0.35 | 1791.48 | 0.0291 |
| multi-replica (8x H20 GPUs) | 32 | 200 | 24.66 | 514.21 | 0.50 | 2221.27 | 0.0198 |

## Deploy Command

### Single replica

```
vllm serve OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5 \
  --omni \
  --host 0.0.0.0 \
  --port 8123 \
  --trust-remote-code \
  --deploy-config vllm_omni/deploy/moss_tts_local.yaml
```

### Multi replica

```
vllm serve OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5 \
  --omni \
  --host 0.0.0.0 \
  --port 8123 \
  --trust-remote-code \
  --deploy-config vllm_omni/deploy/moss_tts_local.yaml \
  --allowed-local-media-path /root/vllm-omni-workspace \
  --stage-overrides '{"0":{"num_replicas":8,"devices":"0"},"1":{"num_replicas":8,"devices":"0"}}' \
  --init-timeout 1800
```

## Benchmark Command

Performance & accuracy

```
SEED_TTS_EVAL_DEVICE=cuda:6 \
SEED_TTS_WER_SAVE_AUDIO_DIR=/root/vllm-omni-workspace/vllm-omni/results/moss_local_debug/refactor \
VLLM_OMNI_BENCH_AUDIO_SAMPLE_RATE=48000 \
VLLM_OMNI_BENCH_AUDIO_CHANNELS=2 \
python benchmarks/tts/bench_tts.py \
  --host 127.0.0.1 \
  --port 8123 \
  --model OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5 \
  --task voice_clone \
  --locale en \
  --dataset-path /root/vllm-omni-workspace/seedtts_testset \
  --num-prompts 50 \
  --concurrency 4 \
  --wer-eval \
  --output-dir /root/vllm-omni-workspace/vllm-omni/results/moss_local_debug/kernel
```

Or only performance:

```
VLLM_OMNI_BENCH_AUDIO_SAMPLE_RATE=48000 \
VLLM_OMNI_BENCH_AUDIO_CHANNELS=2 \
vllm bench serve --omni \
  --host 127.0.0.1 \
  --port 8123 \
  --model OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5 \
  --backend openai-audio-speech \
  --endpoint /v1/audio/speech \
  --dataset-name seed-tts \
  --dataset-path /root/vllm-omni-workspace/seedtts_testset \
  --seed-tts-locale en \
  --num-prompts 200 \
  --num-warmups 4 \
  --max-concurrency 16 \
  --request-rate inf \
  --percentile-metrics ttft,e2el,audio_rtf,audio_ttfp,audio_duration,audio_underrun \
  --save-result \
  --result-dir /root/vllm-omni-workspace/vllm-omni/results/moss_local_stage1_profile_bench
```

## Profile Command

```
VLLM_OMNI_BENCH_AUDIO_SAMPLE_RATE=48000 \
VLLM_OMNI_BENCH_AUDIO_CHANNELS=2 \
vllm bench serve --omni \
  --host 127.0.0.1 \
  --port 8123 \
  --model OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5 \
  --backend openai-audio-speech \
  --endpoint /v1/audio/speech \
  --dataset-name seed-tts \
  --dataset-path /root/vllm-omni-workspace/seedtts_testset \
  --seed-tts-locale en \
  --num-prompts 4 \
  --num-warmups 4 \
  --max-concurrency 4 \
  --request-rate inf \
  --percentile-metrics ttft,e2el,audio_rtf,audio_ttfp,audio_duration,audio_underrun \
  --profile \
  --save-result \
  --result-dir /root/vllm-omni-workspace/vllm-omni/results/moss_local_stage1_profile_bench
```

## Simple request

Streaming:

```
curl -N http://127.0.0.1:8123/v1/audio/speech \
  -H 'Content-Type: application/json' \
  -o /root/vllm-omni-workspace/moss_local_stream.wav \
  -d '{
    "model": "OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5",
    "input": "The rain had been falling since late afternoon, soft at first, then steadily heavier, until the entire city seemed to blur behind a silver curtain. By the time Daniel reached the old train station, the streets were shining like black glass, reflecting the yellow glow of streetlamps and the restless movement of passing cars. He pulled his coat tighter around his shoulders and stepped under the wide iron roof of the platform, where the air smelled of wet stone, engine smoke, and coffee from a small kiosk near the entrance.",
    "ref_audio": "file:///root/vllm-omni-workspace/vllm-omni/local_v15_test.wav",
    "stream": true,
    "stream_format": "audio",
    "response_format": "wav"
  }'
```

Non-streaming:

```
curl -N http://127.0.0.1:8123/v1/audio/speech \
  -H 'Content-Type: application/json' \
  -o /root/vllm-omni-workspace/moss_local_stream.wav \
  -d '{
    "model": "OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5",
    "input": "The rain had been falling since late afternoon, soft at first, then steadily heavier, until the entire city seemed to blur behind a silver curtain. By the time Daniel reached the old train station, the streets were shining like black glass, reflecting the yellow glow of streetlamps and the restless movement of passing cars. He pulled his coat tighter around his shoulders and stepped under the wide iron roof of the platform, where the air smelled of wet stone, engine smoke, and coffee from a small kiosk near the entrance.",
    "ref_audio": "file:///root/vllm-omni-workspace/vllm-omni/local_v15_test.wav",
    "stream": false,
    "response_format": "wav"
  }'
```

**Comment by nagisa-kunhah** [comment]
Hi, I'm working on CUDA graphs for stage 0 and optimizing sampling. I plan to propose this tomorrow. Thanks!

Update: I've profiled and benchmarked the sampling-side optimization under high concurrency, but the gains aren't compelling enough to prioritize. Sampler kernels represent only ~4.5%–4.7% of stage-0 CUDA time, and the Triton sampler path shows limited benefit alongside service-level regressions. I'd propose focusing on the CUDA graph approach in #4938 first, which is simpler and cleaner.

**Comment by Ma1oneZhang** [comment]
Hi, I'd like to take on the `Stage Transfer and Chunking` work. I'll provide an update as soon as I have any progress.

Quick update – it looks like @BruceLoveDecimal has already done a lot of work on the Stage Transfer and Chunking part. Could you share more details on your current progress so we can collaborate more effectively?

**Comment by wuli666** [comment]
Hi, I'd like to try working on the first item under “Serving Configuration and Scheduling”. I'll share updates as soon as I make progress.

**Comment by frank-2077** [comment]
Hi @gcanlin, I'd like to work on the first item under Observability and Warmup. My understanding is that this requires a code PR adding opt-in MOSS-specific profiling/timing for reference encoding, backbone decode, local frame decode, Stage0-to-Stage1 transfer, codec decode, and output formatting—not only a one-off performance report. I plan to reuse the existing transfer metrics, avoid CUDA synchronization in the hot path, add focused tests/docs, and provide warm/cold plus streaming/non-streaming benchmark evidence. Should these timings be visible only in torch/nsys traces, or should they also be exposed through --log-stats and benchmark JSON?

**Comment by skyloevil** [comment]
Taking the single-flight item under **Reference Audio Preprocessing** — draft PR: #5686.

The cache is written only *after* an encode finishes, so it can't dedupe requests arriving while one is in flight. Eight concurrent requests for a cold reference ran eight codec passes; they now run one.

Also: the item above it (content-addressed cache) already looks implemented in `reference_encoder.py` — sha1 content hash, named-voice keys with `created_at`, CPU storage, clone-on-return. Probably safe to check off.

Happy to take the batched reference-audio encoding item next if nobody has it.

**Comment by pxljs** [comment]
Hi @gcanlin, I checked current main after #5197 and #5530.
MossTTSLocalDepthTransformer.generate_frame() still re-prefills the growing 1..n_vq prefix for every codebook, and I could not find an incremental within-frame KV-cache path.
I’d like to work specifically on the RFC item “Reuse local-transformer KV within a frame”, preserving the per-frame cache reset and GPT-J/interleaved RoPE numerics, with eager-vs-cached logits/code equivalence tests and benchmarks. Is anyone already working on this subtask?

**Comment by YuDeng0102** [comment]
Hi @pxljs, I saw your comment—thanks for checking this subtask. I had already started implementing the same within-frame local-transformer KV reuse and have opened draft PR #5903 to make the current work visible. It includes a fixed-address per-frame KV workspace, the incremental path, focused correctness/CUDA graph tests, and preliminary A100 40 GB numbers; real-checkpoint BF16 parity and the post-change end-to-end benchmark are still marked as TODOs in the draft.

Have you already started an implementation? If so, could you share your branch or planned approach? I’m happy to compare designs and coordinate so we avoid duplicating work.

**Comment by pxljs** [comment]
> Hi [@pxljs](https://github.com/pxljs), I saw your comment—thanks for checking this subtask. I had already started implementing the same within-frame local-transformer KV reuse and have opened draft PR [#5903](https://github.com/vllm-project/vllm-omni/pull/5903) to make the current work visible. It includes a fixed-address per-frame KV workspace, the incremental path, focused correctness/CUDA graph tests, and preliminary A100 40 GB numbers; real-checkpoint BF16 parity and the post-change end-to-end benchmark are still marked as TODOs in the draft.
> 
> Have you already started an implementation? If so, could you share your branch or planned approach? I’m happy to compare designs and coordinate so we avoid duplicating work.

ok,thank u,i have not started yet.

**Comment by YuDeng0102** [comment]
Hi @gcanlin, I’ve completed the RFC item “Reuse local-transformer KV within a frame” in #5903.  
The PR adds a checkpoint-gated, fixed-address frame-local KV cache while preserving the existing GPT-J/interleaved RoPE and per-frame reset semantics. It includes eager/compiled logits and codes parity tests, CUDA Graph replay coverage, and warmed C1/C8 end-to-end results on an A100 40 GB.  
The PR is ready for review. I’d appreciate your feedback when you have time. Thanks!
